[MUSIC] Scrape, store and process yt links from TADB for songs in a users local library

the-black-eagle opened the pull request:
## Description
This PR extends the python generic artist scraper to fetch the associated yt video links for an artist from theaudiodb, and then match them up to songs already scanned into the users music library.  This allows skins to then display links to the corresponding song video.

The scraper can fetch all the known video links for each artist from theaudiodb and core will store the links in the db against the appropriate songs.  If the supplied link also includes a thumbnail then that is also stored and available in the art for a song via `ListItem.Art(videothumb)`

The links stored in the db can be updated via json or scraped by the artist scraper.  They can be exported and re-imported via nfo files.

## Motivation and context

This started out as a request on the forum -> https://forum.kodi.tv/showthread.php?tid=367017

## How has this been tested?
Tested with a modified version of confluence initially and then Estuary.  Ensured works for both local and shared MySQL databases.  Tested export and import via NFO files and updating links via json.  There has also been some testing by others which is documented in the linked thread.

## What is the effect on users?
Needs skin support but can allow for video trailers for songs in a similar way to trailers for movies, or a direct link to the yt video, as in the screenshot below.

## Screenshots (if appropriate):

Before re-scraping the artist - 

<a href="https://ibb.co/Nj2NKBk"><img src="https://i.ibb.co/G2QMRS1/screenshot00068.png" alt="screenshot00068" border="0"></a>

After re-scraping - 

<a href="https://ibb.co/w631Jtj"><img src="https://i.ibb.co/YjYGDgw/screenshot00069.png" alt="screenshot00069" border="0"></a>

<a href="https://ibb.co/C95zJ5Q"><img src="https://i.ibb.co/qgMd0Mp/screenshot00067.png" alt="screenshot00067" border="0"></a>

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

the-black-eagle committed:
[MUSIC] Add, store and process scraped video links for artist songs

the-black-eagle committed:
[JSON] Add new song video fields to json-rpc

zag2me commented:
Thanks for the [testbuild](https://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20230201-68b7de89-music_video_links-x64.exe), I did some testing today on this PR:

I've tried with both as a new install and upgrade from v20 nexus. All videos were matched correctly using the artist info scraper.

![Screenshot 2023-02-03 140932](https://user-images.githubusercontent.com/1050512/216634752-46ab13e9-6f13-48d2-8b80-85475cb10244.png)

The only thing I noticed was that re-scanning the music collection did not update the links, I had to entirely remove the source and add again.


Will a follow up PR be added for the Estuary [skin changes](https://www.dropbox.com/s/ybo7brwba1qexp2/skin.estuary.zip?dl=0)?

the-black-eagle commented:
@zag2me Thanks for testing!  Re-scraping an artist and answering YES to the dialog regarding local info (assuming you have artist nfo files) is enough to re-scrape the artist and fetch and store the video links.

There is a python addon in the related forum thread that will either update the links for a single artist given an artistid from the db, or all artists given no artistid.   In the modded version of Estuary there is a `fetch videolinks` button on the artist info page that will run it to fetch them for that particular artist.  Running the addon from program addons will fetch links for all artists. It works but my python is a little rough!

I will not be raising a PR for Estuary.  I didn't for other ListItems that I have added, but (assuming this gets approval), I will document it in the wiki and in the appropriate forum thread.

Once again, thanks for testing.

DaveTBlake commented:
Yet to dig into this, but just a repo/logistics thought: the scraper exists as master in the scraper repo that gets synced into xbmc, so it will need to a PR there for just scarper changes, and then a repo sync and then the changes to xbmc commited. At least that is my understanding of how we manage addons. Shish! 
For review and testing etc. of the feature it is great to have everything here in this PR, just the final commit may need a diffewrent strategy.

the-black-eagle commented:
@DaveTBlake Not a problem to change the strategy for the final commit.  Scraper stuff is all in one commit in any case, which can be dropped from this PR and raised against the scraper repo in the future.

Thanks in advance for any review comments you may have.

enen92 commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -6850,13 +6850,20 @@ const infomap container_str[]  = {{ "property",         CONTAINER_PROPERTY },
-///     <p><hr>
-///     @skinning_v20 **[New Infolabel]** \link ListItem_HdrType `ListItem.HdrType`\endlink
-///   }
+///   \table_row3{   <b>`ListItem.HdrType`</b>,
```
Why is hdr type part of the diff?

enen92 commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -6850,13 +6850,20 @@ const infomap container_str[]  = {{ "property",         CONTAINER_PROPERTY },
+///   \table_row3{   <b>`ListItem.SongVideoURL`</b>,
+///                  \anchor ListItem_SongVideoURL
+///                  _string_,
+///     @return youtube plugin link to a youtube video of a song
```
Maybe better to use video URL and avoid mentioning YouTube, the add-on is responsible for setting the plugin://YouTube URL isn't it?

the-black-eagle replied:
Yeah, the scraper formats up the entire URL when it fetches the link(s).  So the listitem will contain something like `plugin://plugin.video.youtube/?action=play_video&videoid=someYTlink`.  I would have though in that case that the documentation is correct, but I can change it if you'd prefer ?

enen92 replied:
I think it would be better changed. `SongVideoURL` info doesn't need to be necessarily a youtube plugin link just something Kodi can perceive as a video representative of the song - it just happens that youtube is the norm. But could simply be some other video url or a simple http link

the-black-eagle replied:
The artist scraper specifically fetches yt links because that's the only data theaudiodb is holding and it's the only online db that Kodi uses that actually holds any info like that. I understand why you might want it to be more generic as in just a url to somewhere but it can only ever be a youtube link as that is the only url data to fetch.

enen92 replied:
I understand but media providers and addons (scrapers) in this case is not the same as the Kodi core. Kodi can store any string there (or any video to be more precise) its up to the addon to provide a youtube link. You could probably add (e.g. youtube) to the documentation but explicitly state it has to be a youtube link is somehow wrong

enen92 commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -6850,13 +6850,20 @@ const infomap container_str[]  = {{ "property",         CONTAINER_PROPERTY },
+///                  _string_,
+///     @return youtube plugin link to a youtube video of a song
+///     <p><hr>
+///     @skinning_v22 **[New Infolabel]** \link ListItem_SongVideoURL `ListItem.SongVideoURL`\endlink
```
Skinning 21 :)

enen92 commented on xbmc/music/Artist.h:
```diff
@@ -29,6 +29,15 @@ class CDiscoAlbum
+class CArtistVideoLinks
+{
+public:
+  std::string strTitle;
```
Please avoid Hungarian notation. Since all members are public you likely want a struct instead of a class?
If not it's better to create the necessary getters/setters :)

enen92 commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -2180,6 +2224,101 @@ bool CMusicDatabase::ClearArtistLastScrapedTime(int idArtist)
+
+    auto end = std::chrono::steady_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+    CLog::Log(LOGDEBUG, "MusicDatabase::{0}: Time to store videolinks {1}ms query took {2}ms",
```
Use LogF and drop `__function__`

enen92 commented on xbmc/music/MusicDatabase.h:
```diff
@@ -171,6 +173,8 @@ class CMusicDatabase : public CDatabase
               int iBitRate,
               int iSampleRate,
               int iChannels,
+              const std::string& strSongVideoURL,
```
Please avoid Hungarian notation (see code guidelines)

enen92 commented on xbmc/music/MusicDatabase.h:
```diff
@@ -171,6 +173,8 @@ class CMusicDatabase : public CDatabase
               int iSampleRate,
               int iChannels,
+              const std::string& strSongVideoURL,
+              const std::string& strSongVideoThumb,
```
Same

enen92 commented on xbmc/music/Song.h:
```diff
@@ -194,6 +194,8 @@ class CSong final : public ISerializable
   int iChannels;
   std::string strRecordLabel; // Record label from tag for album processing by CMusicInfoScanner::FileItemsToAlbums
   std::string strAlbumType; // (Musicbrainz release type) album type from tag for album processing by CMusicInfoScanner::FileItemsToAlbums
+  std::string strSongVideoURL; // url to song video on youtube
```
Use only songVideoURL without the str prefix, same for the other var below

enen92 commented on xbmc/music/tags/MusicInfoTag.h:
```diff
@@ -255,6 +259,8 @@ class CMusicInfoTag final : public IArchivable, public ISerializable, public ISo
   int m_bitrate;
   std::string m_stationName;
   std::string m_stationArt; // Used to fetch thumb URL for Shoutcasts
+  std::string m_strSongVideoURL; // youtube link to a video for a song
```
Same for these new vars

enen92 reviewed (commented):
LGTM, a few comments left

DaveTBlake commented on xbmc/music/MusicDatabase.h:
```diff
@@ -140,6 +140,8 @@ class CMusicDatabase : public CDatabase
    \param userrating [in] a userrating (my rating) for the song
    \param votes [in] a vote counter for the song rating
+   \param strSongvideoURL [in] url to video of the song
+   \param strSongVideoThumb [in] url to thumb of song video
```
Questioning making the videothumb part of song CRUD. This thumb is just some type of art, and we have deliberately moved away from having fixed parameters for those. I understand it relates to the videoURL (which does belong here), but the prime relationship is some art for the song. So designwise where AddSong gets called it could be followed by an AddArt call rather than having the thumb (some special art) as a parameter of AddSong

DaveTBlake commented:
@the-black-eagle sorry for the long delay. I'm generally happy with how you have implemented this, but unfortunately not had chance to test it (but I know that you will have been thorough). Small niggle over having a type of art as a parameter in AddSong, it seems a backwards step to have any "special" art in even internal calls, but I get why you have done it as you have.

I also wonder if the fetching of video links in GetArtist could be done in query rather the dataset loops, I'll give it some thought later today. If you have a v83 db populated with test data I can play with that will be useful.

the-black-eagle commented:
@DaveTBlake Sure I have a testing DB available, ping me on slack.  Unsure about your comment of fetching videolinks in dataset loops ?  There is a query that does that.  Perhaps you are confusing it with `AddArtistVideoLinks()` which does use two dataset loops.  There is probably a much better way of matching them up.  At the time I went with the simplest way of doing it but I'm aware that it's not optimal and that there is most likely a better way to do it.

@enen92  Think I've addressed everything you raised.  Thanks for the review!

enen92 commented on xbmc/music/tags/MusicInfoTag.h:
```diff
@@ -157,6 +159,8 @@ class CMusicInfoTag final : public IArchivable, public ISerializable, public ISo
   void SetAlbumReleaseStatus(const std::string& strReleaseStatus);
   void SetStationName(const std::string& strStationName); // name of online radio station
   void SetStationArt(const std::string& strStationArt);
+  void SetSongVideoURL(const std::string& strSongVideoURL); // youtube link to video of song
```
same here

enen92 commented on xbmc/music/Song.h:
```diff
@@ -194,6 +194,8 @@ class CSong final : public ISerializable
   int iChannels;
   std::string strRecordLabel; // Record label from tag for album processing by CMusicInfoScanner::FileItemsToAlbums
   std::string strAlbumType; // (Musicbrainz release type) album type from tag for album processing by CMusicInfoScanner::FileItemsToAlbums
+  std::string songVideoURL; // url to song video on youtube
```
more youtubes :)

enen92 reviewed (changes requested):
You change the json-rpc schema definitions so the minor (not the revision!) json-rpc version needs a bump in https://github.com/xbmc/xbmc/blob/29b96fd773c88fdd54e06e0e60413731d6bf018f/xbmc/interfaces/json-rpc/schema/version.txt

Sorry for not noticing this earlier

DaveTBlake commented:
I'm sure there was a JSON API minor version bump in this when I first looked at it.

enen92 commented:
> I'm sure there was a JSON API minor version bump in this when I first looked at it.

Fwiw I would also swear there was the API change there too :) maybe some fuckup with the update

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -2180,6 +2223,101 @@ bool CMusicDatabase::ClearArtistLastScrapedTime(int idArtist)
   return ExecuteQuery(strSQL);
 }
 
+bool CMusicDatabase::AddArtistVideoLinks(const CArtist& artist)
```
Assuming that there will be more songs for any given album artist than there will be videolink it would be better to loop through the links rather than the songs when doing the match up.

Querying the songs for a videolink (when matching on album artist and song title there could be more than one song) with something like 
```
SELECT idSong FROM song 
WHERE strMusicBrainzTrackID = '%s' OR 
(EXISTS (SELECT 1 FROM album_artist WHERE idAlbum = song.idAlbum AND idArtist = %i ) 
AND strTitle LIKE '%%s%%')

```
Then updating song and art tables accordingly.

the-black-eagle replied:
Ah well, matching on one or the other is because some songs can be remixes or re-releases and have the same title but different mbids  E.G. ZZ Top - Tush has an mbid of both `fe8de756-c0b0-468f-9304-e92821f0d2fb` and `699fe8a8-7c7e-44dc-a592-edf307888165`. Both are in my music library but only one is returned from TADB for the music video.  Similarly, their track 'Legs' has at least three different mbids, `b7072e8d-39f8-4e7c-8c30-4317ca590d2e`, `19bac3b7-5452-400c-bf43-54469c501c80` and `23c16131-8304-46b3-a380-67348170fac4`.  Technically, the third is an edited version but the other two are the same track.

However, for the purposes of adding a video link to play a song, I would expect all tracks named 'Legs' by 'ZZ Top' to link to a video if one exists.  The fact that I have a slightly shorter edited version is not the point and in any case, if there was an mbid link to the shorter video (if it even exists) then it would match that first.

So yes, there are good reasons why the matching is done the way it is, otherwise we will just get users saying "I have song x by artist y and there is a link on tadb for a video but Kodi doesn't see it".

I have multiple logs that can show the matching at work if you'd like to peruse them.

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -2180,6 +2223,101 @@ bool CMusicDatabase::ClearArtistLastScrapedTime(int idArtist)
+
+          if (!videoURL.thumbURL.empty())
+          {
+            // already have a thumb for this videoURL ?
```
Existing art could be for a different url, so I would replace comment
`// Already have a videothumb for this song?`

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -2180,6 +2223,101 @@ bool CMusicDatabase::ClearArtistLastScrapedTime(int idArtist)
+          if (!videoURL.thumbURL.empty())
+          {
+            // already have a thumb for this videoURL ?
+            strSQL2 = PrepareSQL("SELECT art_id FROM art "
+                                 "WHERE media_id=%i AND media_type='%s' AND type='videothumb'",
+                                 songId, MediaTypeArtist);
```
Should be `MediaTypeSong` here

DaveTBlake commented on xbmc/interfaces/json-rpc/schema/methods.json:
```diff
@@ -1220,7 +1220,9 @@
-      { "name": "bpm", "$ref": "Optional.Integer" }
+      { "name": "bpm", "$ref": "Optional.Integer" },
+      { "name": "songvideourl", "$ref": "Optional.String" },
+      { "name": "songvideothumb", "$ref": "Optional.String" }
```
Just thinking should a songs videothumb is accessed via JSON like all other song art, no special parameter or field for it.

the-black-eagle commented:
@enen92 Think I have addressed everything you pointed out.  Thanks for the review.
@DaveTBlake Think I have also addressed everything you commented on.  The art is no longer associated with a song other than through the art table, and I have made sure to only use CAPITALS for SQL keywords! ;)

If you both approve, then I'll re-push this one last time, dropping the python scraper stuff from here and raising that against the upstream scraper repo.

enen92 reviewed (approved):
LGTM. Would be nice if you can apply clang-format before going in (eg adding marks for disabling it near the affected region as I think the suggestion might make it less readable)

DaveTBlake commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -2057,7 +2070,7 @@ bool CMusicDatabase::GetArtist(int idArtist, CArtist& artist, bool fetchAll /* =
       return false;
     }
-
+    strSQL2 = strSQL + " - ";
```
Use of strSQL2 sent me expecting use of 2 datasets, then this had me scratching my head for the SQL.
Since it is just text for the debug log statement can you rename it and add an inline comment. Now we no longer have _FUNCTION_ that at least made logs from Windows easier to debug, I have to wonder is soem other text would be better than outputing the  SQL string.

DaveTBlake reviewed (approved):
Looks good @the-black-eagle thanks for the adjustments and your patience. Approving, but a change to that strSQL2 would be cool

manfeed commented:
Please, let me know if there's something I can help with. I mean testing in a skin or the like. I'm eager to see this new feature working in Omega, I really think it can be great. Thanks!

the-black-eagle committed:
[COSMETIC] Add clang-format off/on tags

ksooo commented on xbmc/addons/Scraper.cpp:
```diff
@@ -816,6 +816,20 @@ void DetailsFromFileItem<CArtist>(const CFileItem &item, CArtist &artist)
     artist.discography.emplace_back(discoAlbum);
   }
 
+  int nVideolinks = item.GetProperty("artist.videolinks").asInteger32();
```
1. `const int ...`
2. https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#88-variables => no hungarian notation for variable names, please.

the-black-eagle replied:
Done.

ksooo commented on xbmc/addons/Scraper.cpp:
```diff
@@ -816,6 +816,20 @@ void DetailsFromFileItem<CArtist>(const CFileItem &item, CArtist &artist)
   }
 
+  int nVideolinks = item.GetProperty("artist.videolinks").asInteger32();
+  artist.videolinks.reserve(nVideolinks);
```
I would include this line and the following for loop in `if (nVideoLinks > 0)`- saves some cpu cycles in case of no links available.

the-black-eagle replied:
Done.

ksooo commented on xbmc/addons/Scraper.cpp:
```diff
@@ -816,6 +816,20 @@ void DetailsFromFileItem<CArtist>(const CFileItem &item, CArtist &artist)
+  for (int i = 0; i < nVideolinks; ++i)
+  {
+    std::stringstream prefix;
+    prefix << "artist.videolink" << i + 1;
```
You could avoid this + 1 operation per for loop iteration if you rewrite the for loop `for (int i = 1; i <= nVideolinks; ++i)` if you follow my previous suggestion to check the number of links before entering the loop.

the-black-eagle replied:
Done.

ksooo commented on xbmc/music/Artist.h:
```diff
@@ -29,6 +29,15 @@ class CDiscoAlbum
 
+struct ArtistVideoLinks
+{
+public:
```
No need for `public:`here as this is the default for structs. You can remove this line.

the-black-eagle replied:
Done.

ksooo commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -9288,6 +9415,9 @@ void CMusicDatabase::UpdateTables(int version)
   }
+
+  if (version < 83)
+    m_pDS->exec("ALTER TABLE song ADD strVideoURL TEXT");
```
Maybe you want to add a blank line after this line to increase readability of the code.

the-black-eagle replied:
Done.

ksooo commented on xbmc/music/MusicDatabase.h:
```diff
@@ -140,6 +140,7 @@ class CMusicDatabase : public CDatabase
    \param rating [in] a rating for the song
    \param userrating [in] a userrating (my rating) for the song
    \param votes [in] a vote counter for the song rating
+   \param strSongvideoURL [in] url to video of the song
```
`strSongvideoURL` -> `songvideoURL` to match the actual parameter name

the-black-eagle replied:
Done.

ksooo commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -1238,7 +1242,8 @@ bool CMusicDatabase::UpdateSong(CSong& song, bool bArtists /*= true*/, bool bArt
                           song.rating, song.userrating, song.votes, //
                           song.replayGain, //
-                          song.iBPM, song.iBitRate, song.iSampleRate, song.iChannels);
+                          song.iBPM, song.iBitRate, song.iSampleRate, song.iChannels, //
```
What is the trailing `//` good for?

the-black-eagle replied:
No idea tbh.  I just followed what is already there, as this is the way several functions are written in the music database.

ksooo replied:
I'd say if we see no sense in this, let's remove it.

the-black-eagle replied:
I did some digging and its the second answer here https://stackoverflow.com/questions/24974919/how-to-mark-a-region-so-clang-format-wont-touch-it 

Although I've touched this code and the way it ls laid out would make it a prime candidate for clang to reformat, it isn't in the diff that Jenkins produces.  If I select the block of code in my IDE and apply clang-format to the selected block, with the '//'s at the end, nothing happens.  If I remove them, clang re-formats the selected block.

Because it makes the code much more readable, I'd prefer to leave it as the `clang-format on / off` stuff in GUIInfoManager doesn't appear to make a difference.

ksooo replied:
As you which, I will not fight you on this. 

But if everybody invents own coding guidelines, we can stop using clang-format and maintaining a coding guideline. Tricking our tool that ensures that we obey to our (!) own coding guidelines is imo not the way to go. I clang-format is configured that we want to have certain stuff on one line, there is no justification in using tricks like empty comments.

ksooo commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -2037,11 +2047,14 @@ bool CMusicDatabase::GetArtist(int idArtist, CArtist& artist, bool fetchAll /* =
       return false; // not in the database
 
     std::string strSQL;
+    std::string debugSQL;
```
Remove the declaration here and move it to line 2073, where you actually need this variable.

the-black-eagle replied:
Done.

ksooo commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -2180,6 +2217,93 @@ bool CMusicDatabase::ClearArtistLastScrapedTime(int idArtist)
+  std::string strSQL = PrepareSQL("UPDATE song SET strVideoURL = NULL WHERE idAlbum IN "
+                                  "(SELECT idAlbum FROM album_artist WHERE idArtist = %i)",
+                                  idArtist);
+  bool res = ExecuteQuery(strSQL);
+  if (!res)
+    return res;
```
```c++
if (!ExecuteQuery...)
  return false;
```
... would save you the variable here.

the-black-eagle replied:
Done.

the-black-eagle commented on xbmc/music/MusicDatabase.h:
```diff
@@ -210,6 +212,7 @@ class CMusicDatabase : public CDatabase
    \param iBitRate [in] the bitrate of the song file
    \param iSampleRate [in] the sample rate of the song file
    \param iChannels [in] the number of audio channels in the song file
+   \param strSongVideoURL [in] url link to a video of the song
```
Done.

ksooo commented on xbmc/addons/Scraper.cpp:
```diff
@@ -816,6 +816,23 @@ void DetailsFromFileItem<CArtist>(const CFileItem &item, CArtist &artist)
+  if (numvideolinks > 0)
+  {
+    artist.videolinks.reserve(numvideolinks);
+    for (int i = 1; i < numvideolinks; ++i)
```
`i <= numvideolinks` - otherwise you will loose last link!

the-black-eagle replied:
Thanks!

ksooo commented on xbmc/music/tags/MusicInfoTag.h:
```diff
@@ -157,6 +158,7 @@ class CMusicInfoTag final : public IArchivable, public ISerializable, public ISo
   void SetAlbumReleaseStatus(const std::string& strReleaseStatus);
   void SetStationName(const std::string& strStationName); // name of online radio station
   void SetStationArt(const std::string& strStationArt);
+  void SetSongVideoURL(const std::string& strSongVideoURL); // link to video of song
```
Avoid Hungarian notation: `strSongVideoURL` -> `songVideoURL`

the-black-eagle replied:
Done.

enen92 commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -2077,12 +2090,36 @@ bool CMusicDatabase::GetArtist(int idArtist, CArtist& artist, bool fetchAll /* =
-    CLog::Log(LOGDEBUG, LOGDATABASE, "{0}({1}) - took {2} ms", __FUNCTION__, strSQL,
-              duration.count());
-
+    CLog::LogF(LOGDEBUG, "MusicDatabase::GetArtist - {} - took {} ms", debugSQL, duration.count());
```
LogF already logs the method name

the-black-eagle replied:
Sure.  Just used to writing what I'd like to see in any logs.

DaveTBlake commented on xbmc/interfaces/json-rpc/schema/version.txt:
```diff
@@ -1 +1 @@
-JSONRPC_VERSION 13.2.1
+JSONRPC_VERSION 13.3.1
```
Bump should be to 13.3.0

the-black-eagle replied:
Yeah, I wasn't too sure although it seemed logical.  Fixed it.  Thanks.

the-black-eagle commented:
@DaveTBlake , @enen92 , @ksooo   If there are no other issues, I'd like to merge this later.  Thank you all for your input.

jurialmunkey commented:
@the-black-eagle - Is there a InfoTagMusic setter in the Python API for this infolabel, or some other way a plugin can set it for a listitem? As far as I can tell, there isn't a way for a plugin to set it, correct?

the-black-eagle commented:
> Is there a InfoTagMusic setter in the Python API for this infolabel, or some other way a plugin can set it for a listitem? As far as I can tell, there isn't a way for a plugin to set it, correct?

No, there isn't anything in the Python API for it.  I can take a look though to see if I can add it.
